akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1037,143 +1045,79 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
         val previous = asyncCallbacksInProgress.get()
-        if (previous != null) {
-          val updated = previous.updated(this, previous(this) + promise)
+        if (previous != null) { // not stopped
+          val updated = promise :: previous
           if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) addToWaiting()
           else true
-        } else {
+        } else // logic was already stopped
           false
-        }
       }
 
-      if (addToWaiting())
-        invokeWithPromise(event, promise).future
-      else
+      if (addToWaiting()) {
+        invokeWithPromise(event, promise)
+        promise.future
+      } else
         Future.failed(streamDetatchedException)
     }
 
-    // removes the promise from the callbacks in promise on complete, called from onComplete
-    private def onFeedbackCompleted(promise: Promise[Done]): Unit = {
-      @tailrec
-      def removeFromWaiting(): Unit = {
-        val previous = asyncCallbacksInProgress.get()
-        if (previous != null) {
-          val newSet = previous(this) - promise
-          val updated =
-            if (newSet.isEmpty) previous - this // no outstanding promises, remove stage from map to avoid leak
-            else previous.updated(this, newSet)
-          if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) removeFromWaiting()
-        }
-      }
-      removeFromWaiting()
-    }
+    //external call
+    override def invoke(event: T): Unit = invokeWithPromise(event, NoPromise)
 
     @tailrec
-    private def invokeWithPromise(event: T, promise: Promise[Done]): Promise[Done] =
+    private def invokeWithPromise(event: T, promise: Promise[Done]): Unit =
       currentState.get() match {
-        // started - can just send message to stream
-        case Initialized ⇒ sendEvent(event, promise)
-        // not started yet
-        case list @ Pending(_) ⇒
-          if (!currentState.compareAndSet(list, Pending(Event(event, OptionVal(promise)) :: list.pendingEvents)))
-            invokeWithPromise(event, promise) // atomicity is failed - try again
-          else promise
-        // initializing is in progress in another thread (initializing thread is managed by Akka)
-        case Initializing ⇒ if (!currentState.compareAndSet(Initializing, Pending(Event(event, OptionVal(promise)) :: Nil))) {
-          (currentState.get(): @unchecked) match {
-            case Pending(_)  ⇒ invokeWithPromise(event, promise) // atomicity is failed - try again
-            case Initialized ⇒ sendEvent(event, promise)
-          }
-        } else promise
+        case Initialized ⇒
+          // started - can just dispatch async message to interpreter
+          onAsyncInput(event, promise)
+
+          // checking `stopped` is an extra volatile read which is likely false, so we check it only after dispatching
+          // the async callback
+          if ((promise ne NoPromise) && stopped) promise.tryFailure(streamDetatchedException)
```

Comment:
Wondering if we actually even need this.

If the stage was stopped when `invokeWithFeedback` tried to add to `asyncCallbacksInProgress` that would already return a failed future, and if it got through there it should be taken care of the fail-promises-on-stop logic. Am I missing a scenario?

Reply:
Good point. I didn't add it but it got there by manually inlining the redundant extra methods (`sendEvent` and `failPromiseOnComplete`). Probably it was just to hard to see if it is necessary or not. I agree that it doesn't seem to be necessary any more.

Reply:
Removed it.